lib/syslog/pool/pool.go

```diff
@@ -0,0 +1,79 @@
+package pool
+
+import (
+	"io"
+)
+
+// A LimitedConnPool is a connection pool, with the property that
+// the total number of live connections is limited to the size
+// parameter passed to New.
+type LimitedConnPool struct {
+	conns  chan io.WriteCloser // Available connections in the pool
+	live   chan struct{}       // Keep a count of living connections (in our hands, or the client)
+	signal chan struct{}       // Used to wake up the connection producer
+	err    chan error          // Send dial errors back to the client
+}
+
+// New returns a new LimitedConnPool.
+func New(dial func() (io.WriteCloser, error), size int) (*LimitedConnPool, error) {
+	// Tentative first try - if this doesn't work, we assume it never will
+	// and fail to initialize.
```

Comment:
I see the value in this but it's a bit risky, I feel like we should try to have a great story around error handling, and not make assumptions here when creating the pool. It could be that the server it's trying to connect to isn't up yet but the program knows it will be at some point in a near future (in a test setup for example).

Reply:
Good point. If we were to remove the fallback addresses in syslog,

```Go
	if len(config.Address) != 0 {
		addropts = []string{config.Address}
	} else if strings.HasPrefix(config.Network, "unix") {
		// This was copied from the standard log/syslog package, they do the same
		// and try to guess at runtime which socket syslogd is using.
		addropts = []string{"/dev/log", "/var/run/syslog", "/var/run/log"}
	} else {
		// The config doesn't point to a unix domain socket, falling back to trying
		// to connect to syslogd over a network interface.
		addropts = []string{"localhost:514"}
	}
```

then it would be simple to remove this early failure. But we currently rely on these dials to fail immediately in order to cycle through all the defaults until we find one that works. I'm also worried that removing these fallbacks could break someone's code.

Reply:
Eh... I see. Let's comment the code explaining why we need to do it this way then, so nobody is surprised about the hacky implementation later on.

Reply:
Done